src/commands/deck.ts

```diff
@@ -285,6 +302,31 @@ export class DeckCommand extends Command {
 		this.logger.info(serializeCommand(interaction), error);
 	}
 
+	protected async upload(filename: string, deck: Buffer): Promise<void> {
+		// with no credentials, we can't upload
+		if (process.env.FTP_URL === undefined) {
+			throw new Error(t`FTP credentials are undefined!`);
+		}
+		// FTP_URL format in .env is ftp://user:password@host:port
+		const ftpUrl = new URL(process.env.FTP_URL);
+
+		// if the specified port is the default for the protocol, i.e. 21,
+		// URL#port returns the empty string, so we need to step in manually
+		let port = parseInt(ftpUrl.port);
+		if (isNaN(port)) {
+			port = 21;
+		}
+		await ftp.access({
+			host: ftpUrl.hostname,
+			port,
+			user: decodeURIComponent(ftpUrl.username), // URL class percent-encodes stuff we don't want it to
+			password: ftpUrl.password,
+			secure: false // current FTP destination does not support SFTP
+		});
```

Comment:
I think you can do the work to generate the ftp.access parameters once in the constructor and store the parameters. When the constructor is run, only warn if FTP_URL isn't set, and leave the parameters null. Then here in upload, you can continue to throw the error if the parameters are null